node/src/main/scala/coop/rchain/node/api/WebApi.scala

```diff
@@ -335,32 +331,71 @@ object WebApi {
+  /**
+    * Private definition of [[Monoid]] for [[Par]] with the implementation of
+    * combining fields used by WebApi representation of Par [[RhoExpr]].
+    */
+  implicit private object ParMonoid extends Monoid[Par] {
+    override def empty: Par = VectorPar()
+
+    override def combine(x: Par, y: Par): Par =
+      x.copy(
+        x.sends ++ y.sends,
+        x.receives ++ y.receives,
+        x.news ++ y.news,
+        x.exprs ++ y.exprs,
+        x.matches ++ y.matches,
+        x.unforgeables ++ y.unforgeables,
+        x.bundles ++ y.bundles,
+        x.connectives ++ y.connectives
+      )
+  }
+
+  /**
+    * Converts [[RhoExpr]] WebApi representation to protobuf [[Par]] type.
+    */
+  def rhoExprToParProto(exp: RhoExpr): Par = exp match {
+    // Nested expressions (Par, Tuple, List and Set are converted to JSON list)
+    case ExprPar(data)   => data.map(rhoExprToParProto).combineAll
+    case ExprTuple(data) => TupleN(data.map(rhoExprToParProto))
+    case ExprList(data)  => List(data.map(rhoExprToParProto))
+    case ExprSet(data)   => Set(data.map(rhoExprToParProto).toSeq)
+    case ExprMap(data)   => Map(data.map { case (k, v) => (String(k), rhoExprToParProto(v)) })
+    // Terminal expressions (here is the data)
+    case ExprBool(data)   => Boolean(data)
+    case ExprInt(data)    => Number(data)
+    case ExprString(data) => String(data)
+    case ExprUri(data)    => Uri(data)
+    // Binary data is decoded from base16 string
+    case ExprBytes(data)  => ByteArray(data.unsafeHexToByteString.toByteArray)
+    case ExprUnforg(data) => unforgToParProto(data)
+  }
+
+  private def unforgToParProto(unforg: RhoUnforg): Par = unforg match {
+    case UnforgPrivate(name)  => Name(name.unsafeHexToByteString)
+    case UnforgDeploy(name)   => DeployId(name.unsafeDecodeHex)
+    case UnforgDeployer(name) => DeployerId(name.unsafeDecodeHex)
   }
 
   // Data request/response protobuf wrappers
 
-  private def toPar(req: DataAtNameRequest): Par =
-    Par(unforgeables = Seq(GUnforgeable(unforgToUnforgProto(req.name))))
+  private def toPar(req: DataAtNameRequest): Par = unforgToParProto(req.name)
 
-  private def toPar(req: DataAtNameByBlockHashRequest): Par =
-    Par(unforgeables = Seq(GUnforgeable(unforgToUnforgProto(req.name))))
+  private def toPar(req: DataAtNameByBlockHashRequest): Par = rhoExprToParProto(req.name)
 
   private def toDataAtNameResponse(req: (Seq[DataWithBlockInfo], Int)): DataAtNameResponse = {
     val (dbs, length) = req
-    val exprsWithBlock = dbs.foldLeft(List[RhoExprWithBlock]()) { (acc, data) =>
-      val exprs = data.postBlockData.flatMap(exprFromParProto)
-      // Implements semantic of Par with Unit: P | Nil ==> P
-      val expr  = if (exprs.size == 1) exprs.head else ExprPar(exprs.toList)
-      val block = data.block
-      RhoExprWithBlock(expr, block) +: acc
+    val exprsWithBlock = dbs.foldLeft(collection.immutable.List[RhoExprWithBlock]()) {
```

Comment:
Does the full namespace for List is needed? It seems to work fine on other places.

Reply:
It is conflicted with `List` from RhoType.scala

https://github.com/rchain/rchain/blob/6ad914c91e5a9c499c5736f651ceffb3ed7ac147/rholang/src/main/scala/coop/rchain/rholang/interpreter/RhoType.scala#L107

But I don't know why it is works in other cases e.g. here

https://github.com/rchain/rchain/blob/28db3868ff83703c9dad18f69db8ce9baeba8c8b/node/src/main/scala/coop/rchain/node/api/WebApi.scala#L37

Reply:
Interesting. We should definitely plan to change RhoType with names not in conflict with built-in types.